nym-vpn-core/crates/nym-vpn-network-config/src/refresh.rs

```diff
@@ -1,47 +1,67 @@
 // Copyright 2024 - Nym Technologies SA <contact@nymtech.net>
 // SPDX-License-Identifier: GPL-3.0-only
 
-use std::{
-    path::{Path, PathBuf},
-    time::Duration,
-};
+use std::{path::PathBuf, time::Duration};
 
 use tokio::task::JoinHandle;
 use tokio_util::sync::CancellationToken;
 
 use super::discovery::Discovery;
 
-fn refresh_discovery_file(config_dir: &Path, network_name: &str) -> anyhow::Result<()> {
-    if !Discovery::path_is_stale(config_dir, network_name)? {
-        return Ok(());
-    }
-    Discovery::fetch(network_name)?.write_to_file(config_dir)?;
-    Ok(())
-}
-
-// Ideally we only refresh the discovery file when the tunnel is up
-#[allow(unused)]
-pub(crate) async fn start_background_discovery_refresh(
+struct DiscoveryRefresher {
     config_path: PathBuf,
     network_name: String,
     cancel_token: CancellationToken,
-) -> JoinHandle<()> {
-    tokio::spawn(async move {
+}
+
+impl DiscoveryRefresher {
+    fn new(config_path: PathBuf, network_name: String, cancel_token: CancellationToken) -> Self {
+        Self {
+            config_path,
+            network_name,
+            cancel_token,
+        }
+    }
+
+    async fn refresh_discovery_file(&self) -> anyhow::Result<()> {
+        tracing::info!("Checking descovery file age");
+        if !Discovery::path_is_stale(self.config_path.as_path(), &self.network_name)? {
+            return Ok(());
+        }
+        tracing::info!("Refreshing and updating file");
+        Discovery::fetch(&self.network_name)
+            .await?
+            .write_to_file(self.config_path.as_path())?;
+        Ok(())
+    }
+
+    async fn run(self) {
         // Check once an hour
-        let mut interval = tokio::time::interval(Duration::from_secs(60 * 60));
+        let mut interval = tokio::time::interval(Duration::from_secs(60));
         interval.tick().await; // initial tick
 
         loop {
             tokio::select! {
                 _ = interval.tick() => {
-                    if let Err(err) = refresh_discovery_file(&config_path, &network_name) {
+                    if let Err(err) = self.refresh_discovery_file().await {
```

Comment:
We probably need to wrap this in a mechanism to abort if stuck, just so we don't end up in frozen state on shutdown. Maybe `self.cancel_token.run_until_cancelled(..)` but we also don't want to forcefully cancel if not necessary. In case we cancel while writing a file

Reply:
In this case the fs won't be cancelled mid write because it is not async. The only thing that can get cancelled is the network `fetch`, which I would say in this context, is cancel safe, as the operation is just a GET.